File: android/app/src/main/java/com/lichso/app/domain/DayInfoProvider.kt
package com.lichso.app.domain

import com.lichso.app.domain.model.*
import com.lichso.app.util.*
import java.time.LocalDate
import java.time.LocalTime
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Provides complete day information by composing all calculators
 */
@Singleton
class DayInfoProvider @Inject constructor() {

    fun getDayInfo(dd: Int, mm: Int, yy: Int): DayInfo {
        val jd = LunarCalendarUtil.jdFromDate(dd, mm, yy)
        val lunar = LunarCalendarUtil.convertSolar2Lunar(dd, mm, yy)

        val yearCanChi = CanChiCalculator.getYearCanChi(lunar.lunarYear)
        val monthCanChi = CanChiCalculator.getMonthCanChi(lunar.lunarMonth, lunar.lunarYear)
        val dayCanChi = CanChiCalculator.getDayCanChi(jd)
        val dayOfWeek = CanChiCalculator.getDayOfWeek(jd)
        val dayOfWeekIndex = CanChiCalculator.getDayOfWeekIndex(jd)

        val moonPhase = MoonPhaseCalculator.getMoonPhase(dd, mm, yy)
        val gioHD = GioHoangDaoCalculator.getGioHoangDao(jd)
        val activities = DayActivityCalculator.getDayActivities(jd, lunar.lunarDay, lunar.lunarMonth)
        val huong = DayActivityCalculator.getHuongTot(jd)
        val solarHoliday = HolidayUtil.getSolarHoliday(dd, mm)
        val lunarHoliday = HolidayUtil.getLunarHoliday(lunar.lunarDay, lunar.lunarMonth)
        val tietKhi = TietKhiCalculator.getCurrentSolarTerm(dd, mm, yy)
        val trucNgay = TrucNgayCalculator.getTrucNgay(jd, lunar.lunarMonth)
        val saoChieu = SaoChieuCalculator.getSaoChieu(jd)

        // Hour Can Chi (current hour)
        val currentHour = LocalTime.now().hour
        val hourCanChi = getHourCanChi(jd, currentHour)

        // Day rating
        val dayRating = calculateDayRating(trucNgay, saoChieu, activities)

        val lunarMonthName = if (lunar.lunarLeap == 1) "Tháng ${lunar.lunarMonth} nhuận" else "Tháng ${lunar.lunarMonth}"

        return DayInfo(
            solar = SolarDate(dd, mm, yy),
            lunar = LunarDate(lunar.lunarDay, lunar.lunarMonth, lunar.lunarYear, lunar.lunarLeap, lunarMonthName),
            jd = jd,
            dayOfWeek = dayOfWeek,
            dayOfWeekIndex = dayOfWeekIndex,
            yearCanChi = yearCanChi,
            monthCanChi = monthCanChi,
            dayCanChi = dayCanChi,
            hourCanChi = hourCanChi,
            moonPhase = MoonPhaseInfo(moonPhase.icon, moonPhase.name, moonPhase.age),
            gioHoangDao = gioHD.map { GioHoangDaoInfo(it.name, it.time) },
            activities = DayActivitiesInfo(activities.nenLam, activities.khongNen, activities.isXauDay, activities.isNguyetKy, activities.isTamNuong),
            huong = HuongTotInfo(huong.thanTai, huong.hyThan, huong.hungThan),
            solarHoliday = solarHoliday,
            lunarHoliday = lunarHoliday,
            tietKhi = TietKhiInfo(
                currentName = tietKhi.current?.name,
                nextName = tietKhi.next?.name,
                nextDd = tietKhi.next?.dd,
                nextMm = tietKhi.next?.mm,
                daysUntilNext = tietKhi.daysUntilNext
            ),
            trucNgay = TrucNgayInfo(trucNgay.name, trucNgay.rating),
            saoChieu = SaoChieuInfo(saoChieu.name, saoChieu.rating),
            dayRating = dayRating,
            isRam = lunar.lunarDay == 15,
            isMung1 = lunar.lunarDay == 1
        )
    }

    fun getCalendarDays(year: Int, month: Int, weekStartSunday: Boolean = false): List<CalendarDay> {
        // month is 1-indexed
        val today = LocalDate.now()
        val firstDay = LocalDate.of(year, month, 1)

        // dayOfWeek.value: 1=Mon..7=Sun
        val offset = if (weekStartSunday) {
            // Sunday start: Sun=0, Mon=1 .. Sat=6
            firstDay.dayOfWeek.value % 7
        } else {
            // Monday start: Mon=0, Tue=1 .. Sun=6
            firstDay.dayOfWeek.value - 1
        }

        val daysInMonth = firstDay.lengthOfMonth()
        val prevMonth = firstDay.minusMonths(1)
        val daysInPrevMonth = prevMonth.lengthOfMonth()

        val days = mutableListOf<CalendarDay>()

        // Previous month days
        for (i in offset - 1 downTo 0) {
            val d = daysInPrevMonth - i
            val pm = prevMonth.monthValue
            val py = prevMonth.year
            val lunar = LunarCalendarUtil.convertSolar2Lunar(d, pm, py)
            val lunStr = if (lunar.lunarDay == 1) "${lunar.lunarDay}/${lunar.lunarMonth}" else "${lunar.lunarDay}"
            val dow = LocalDate.of(py, pm, d).dayOfWeek
            days.add(
                CalendarDay(
                    solarDay = d, solarMonth = pm, solarYear = py,
                    lunarDay = lunar.lunarDay, lunarMonth = lunar.lunarMonth,
                    isCurrentMonth = false, isToday = false,
                    isSunday = dow.value == 7, isSaturday = dow.value == 6,
                    isHoliday = false, hasEvent = false, lunarDisplayText = lunStr
                )
            )
        }

        // Current month days
        for (d in 1..daysInMonth) {
            val date = LocalDate.of(year, month, d)
            val lunar = LunarCalendarUtil.convertSolar2Lunar(d, month, year)
            val lunStr = if (lunar.lunarDay == 1) "${lunar.lunarDay}/${lunar.lunarMonth}" else "${lunar.lunarDay}"
            val dow = date.dayOfWeek
            val sHol = HolidayUtil.getSolarHoliday(d, month)
            val lHol = HolidayUtil.getLunarHoliday(lunar.lunarDay, lunar.lunarMonth)
            val hasEvent = lunar.lunarDay == 1 || lunar.lunarDay == 15 || sHol != null || lHol != null

            // Compute day rating for hoang dao marking
            val jd = LunarCalendarUtil.jdFromDate(d, month, year)
            val trucNgay = TrucNgayCalculator.getTrucNgay(jd, lunar.lunarMonth)
            val saoChieu = SaoChieuCalculator.getSaoChieu(jd)
            val activities = DayActivityCalculator.getDayActivities(jd, lunar.lunarDay, lunar.lunarMonth)
            val rating = calculateDayRating(trucNgay, saoChieu, activities)

            days.add(
                CalendarDay(
                    solarDay = d, solarMonth = month, solarYear = year,
                    lunarDay = lunar.lunarDay, lunarMonth = lunar.lunarMonth,
                    isCurrentMonth = true,
                    isToday = date == today,
                    isSunday = dow.value == 7,
                    isSaturday = dow.value == 6,
                    isHoliday = sHol != null || lHol != null,
                    hasEvent = hasEvent,
                    lunarDisplayText = lunStr,
                    dayRatingLabel = rating.label
                )
            )
        }

        // Next month days
        val totalCells = offset + daysInMonth
        val remaining = if (totalCells % 7 == 0) 0 else 7 - (totalCells % 7)
        val nextMonth = firstDay.plusMonths(1)
        for (d in 1..remaining) {
            val nm = nextMonth.monthValue
            val ny = nextMonth.year
            val lunar = LunarCalendarUtil.convertSolar2Lunar(d, nm, ny)
            val lunStr = if (lunar.lunarDay == 1) "${lunar.lunarDay}/${lunar.lunarMonth}" else "${lunar.lunarDay}"
            val dow = LocalDate.of(ny, nm, d).dayOfWeek
            days.add(
                CalendarDay(
                    solarDay = d, solarMonth = nm, solarYear = ny,
                    lunarDay = lunar.lunarDay, lunarMonth = lunar.lunarMonth,
                    isCurrentMonth = false, isToday = false,
                    isSunday = dow.value == 7, isSaturday = dow.value == 6,
                    isHoliday = false, hasEvent = false, lunarDisplayText = lunStr
                )
            )
        }

        return days
    }

    fun getUpcomingEvents(dd: Int, mm: Int, yy: Int): List<UpcomingEvent> {
        val events = mutableListOf<UpcomingEvent>()
        val todayJd = LunarCalendarUtil.jdFromDate(dd, mm, yy)

        for (i in 0 until 14) {
            val checkJd = todayJd + i
            val (cd, cm, cy) = LunarCalendarUtil.jdToDate(checkJd)
            val lunar = LunarCalendarUtil.convertSolar2Lunar(cd, cm, cy)
            val timeStr = when (i) {
                0 -> "Hôm nay"
                1 -> "Ngày mai · $cd/$cm"
                else -> "$cd/$cm"
            }

            if (lunar.lunarDay == 15) {
                events.add(UpcomingEvent("Rằm tháng ${lunar.lunarMonth} Âm lịch", timeStr, "Âm lịch", EventColor.TEAL))
            }
            if (lunar.lunarDay == 1) {
                events.add(UpcomingEvent("Mùng 1 tháng ${lunar.lunarMonth} Âm lịch", timeStr, "Âm lịch", EventColor.TEAL))
            }
            val sHol = HolidayUtil.getSolarHoliday(cd, cm)
            if (sHol != null) {
                events.add(UpcomingEvent(sHol, timeStr, "Ngày lễ", EventColor.RED))
            }
            val lHol = HolidayUtil.getLunarHoliday(lunar.lunarDay, lunar.lunarMonth)
            if (lHol != null && lunar.lunarDay != 15 && lunar.lunarDay != 1) {
                events.add(UpcomingEvent(lHol, timeStr, "Âm lịch", EventColor.GOLD))
            }

            if (events.size >= 5) break
        }

        // Add upcoming solar term
        val tkInfo = TietKhiCalculator.getCurrentSolarTerm(dd, mm, yy)
        if (tkInfo.next != null && tkInfo.daysUntilNext <= 14) {
            val nd = tkInfo.next
            val timeStr = when (tkInfo.daysUntilNext) {
                0 -> "Hôm nay"
                1 -> "Ngày mai · ${nd.dd}/${nd.mm}"
                else -> "${nd.dd}/${nd.mm} · Bắt đầu tiết khí mới"
            }
            events.add(UpcomingEvent("Tiết ${nd.name}", timeStr, "Tiết khí", EventColor.RED))
        }

        return events.take(4)
    }

    private fun getHourCanChi(jd: Int, hour: Int): String {
        val can = CanChiCalculator.THIEN_CAN
        val chi = CanChiCalculator.DIA_CHI
        val chiIdx = when (hour) {
            23, 0 -> 0; 1, 2 -> 1; 3, 4 -> 2; 5, 6 -> 3
            7, 8 -> 4; 9, 10 -> 5; 11, 12 -> 6; 13, 14 -> 7
            15, 16 -> 8; 17, 18 -> 9; 19, 20 -> 10; 21, 22 -> 11
            else -> 0
        }
        val dayCan = (jd + 9) % 10
        val canIdx = (dayCan * 2 + chiIdx) % 10
        return "${can[canIdx]} ${chi[chiIdx]}"
    }

    private fun calculateDayRating(
        truc: TrucNgayCalculator.TrucNgayInfo,
        sao: SaoChieuCalculator.SaoChieuInfo,
        activities: DayActivityCalculator.DayActivities
    ): DayRatingInfo {
        var score = 50

        // Trực ngày rating contribution
        when (truc.rating) {
            "Tốt" -> score += 20
            "Xấu" -> score -= 15
        }

        // Sao chiếu rating contribution
        when (sao.rating) {
            "Tốt" -> score += 20
            "Xấu" -> score -= 15
        }

        // Nguyệt kỵ / Tam nương penalty
        if (activities.isNguyetKy) score -= 15
        if (activities.isTamNuong) score -= 10

        // Base bonus from good activities
        score += (activities.nenLam.size * 2).coerceAtMost(10)

        score = score.coerceIn(10, 100)

        val label = when {
            score >= 80 -> "Rất tốt"
            score >= 60 -> "Tốt"
            score >= 40 -> "Trung bình"
            else -> "Xấu"
        }

        return DayRatingInfo(label, score)
    }
}

File: ios/LichSo/Domain/DayInfoProvider.swift
import Foundation

// MARK: - Provides complete day information by composing all calculators

class DayInfoProvider {
    static let shared = DayInfoProvider()

    func getDayInfo(dd: Int, mm: Int, yy: Int) -> DayInfo {
        let jd = LunarCalendarUtil.jdFromDate(dd: dd, mm: mm, yy: yy)
        let lunar = LunarCalendarUtil.convertSolar2Lunar(dd: dd, mm: mm, yy: yy)

        let yearCanChi = CanChiCalculator.getYearCanChi(lunarYear: lunar.lunarYear)
        let monthCanChi = CanChiCalculator.getMonthCanChi(lunarMonth: lunar.lunarMonth, lunarYear: lunar.lunarYear)
        let dayCanChi = CanChiCalculator.getDayCanChi(jd: jd)
        let dayOfWeek = CanChiCalculator.getDayOfWeek(jd: jd)
        let dayOfWeekIndex = CanChiCalculator.getDayOfWeekIndex(jd: jd)

        let moonPhase = MoonPhaseCalculator.getMoonPhase(dd: dd, mm: mm, yy: yy)
        let gioHD = GioHoangDaoCalculator.getGioHoangDao(jd: jd)
        let activities = DayActivityCalculator.getDayActivities(jd: jd, lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth)
        let huong = DayActivityCalculator.getHuongTot(jd: jd)
        let solarHoliday = HolidayUtil.getSolarHoliday(dd: dd, mm: mm)
        let lunarHoliday = HolidayUtil.getLunarHoliday(lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth)
        let tietKhi = TietKhiCalculator.getCurrentSolarTerm(dd: dd, mm: mm, yy: yy)
        let trucNgay = TrucNgayCalculator.getTrucNgay(jd: jd, lunarMonth: lunar.lunarMonth)
        let saoChieu = SaoChieuCalculator.getSaoChieu(jd: jd)

        // Hour Can Chi (current hour)
        let currentHour = Calendar.current.component(.hour, from: Date())
        let hourCanChi = getHourCanChi(jd: jd, hour: currentHour)

        // Day rating
        let dayRating = calculateDayRating(truc: trucNgay, sao: saoChieu, activities: activities)

        let lunarMonthName = lunar.lunarLeap == 1 ? "Tháng \(lunar.lunarMonth) nhuận" : "Tháng \(lunar.lunarMonth)"

        return DayInfo(
            solar: SolarDate(dd: dd, mm: mm, yy: yy),
            lunar: LunarDate(day: lunar.lunarDay, month: lunar.lunarMonth, year: lunar.lunarYear, leap: lunar.lunarLeap, monthName: lunarMonthName),
            jd: jd,
            dayOfWeek: dayOfWeek,
            dayOfWeekIndex: dayOfWeekIndex,
            yearCanChi: yearCanChi,
            monthCanChi: monthCanChi,
            dayCanChi: dayCanChi,
            hourCanChi: hourCanChi,
            moonPhase: MoonPhaseInfo(icon: moonPhase.icon, name: moonPhase.name, age: moonPhase.age),
            gioHoangDao: gioHD.map { GioHoangDaoInfo(name: $0.name, time: $0.time) },
            activities: DayActivitiesInfo(nenLam: activities.nenLam, khongNen: activities.khongNen, isXauDay: activities.isXauDay, isNguyetKy: activities.isNguyetKy, isTamNuong: activities.isTamNuong),
            huong: HuongTotInfo(thanTai: huong.thanTai, hyThan: huong.hyThan, hungThan: huong.hungThan),
            solarHoliday: solarHoliday,
            lunarHoliday: lunarHoliday,
            tietKhiInfo: TietKhiInfoModel(
                currentName: tietKhi.current?.name,
                nextName: tietKhi.next?.name,
                nextDd: tietKhi.next?.dd,
                nextMm: tietKhi.next?.mm,
                daysUntilNext: tietKhi.daysUntilNext
            ),
            trucNgayInfo: TrucNgayInfoModel(name: trucNgay.name, rating: trucNgay.rating),
            saoChieuInfo: SaoChieuInfoModel(name: saoChieu.name, rating: saoChieu.rating),
            dayRating: dayRating,
            isRam: lunar.lunarDay == 15,
            isMung1: lunar.lunarDay == 1
        )
    }

    func getCalendarDays(year: Int, month: Int, weekStartSunday: Bool = false) -> [CalendarDay] {
        let today = Date()
        let calendar = Calendar.current
        let todayComponents = calendar.dateComponents([.year, .month, .day], from: today)

        guard let firstDay = calendar.date(from: DateComponents(year: year, month: month, day: 1)) else { return [] }
        let weekday = calendar.component(.weekday, from: firstDay) // 1=Sun..7=Sat

        let offset: Int
        if weekStartSunday {
            offset = weekday - 1
        } else {
            offset = (weekday + 5) % 7 // Mon=0..Sun=6
        }

        let daysInMonth = calendar.range(of: .day, in: .month, for: firstDay)!.count
        let prevMonth = calendar.date(byAdding: .month, value: -1, to: firstDay)!
        let daysInPrevMonth = calendar.range(of: .day, in: .month, for: prevMonth)!.count
        let prevMonthValue = calendar.component(.month, from: prevMonth)
        let prevYearValue = calendar.component(.year, from: prevMonth)

        var days: [CalendarDay] = []

        // Previous month days
        for i in stride(from: offset - 1, through: 0, by: -1) {
            let d = daysInPrevMonth - i
            let lunar = LunarCalendarUtil.convertSolar2Lunar(dd: d, mm: prevMonthValue, yy: prevYearValue)
            let lunStr = lunar.lunarDay == 1 ? "\(lunar.lunarDay)/\(lunar.lunarMonth)" : "\(lunar.lunarDay)"
            let date = calendar.date(from: DateComponents(year: prevYearValue, month: prevMonthValue, day: d))!
            let wd = calendar.component(.weekday, from: date)
            days.append(CalendarDay(
                solarDay: d, solarMonth: prevMonthValue, solarYear: prevYearValue,
                lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth,
                isCurrentMonth: false, isToday: false,
                isSunday: wd == 1, isSaturday: wd == 7,
                isHoliday: false, hasEvent: false, lunarDisplayText: lunStr
            ))
        }

        // Current month days
        for d in 1...daysInMonth {
            let date = calendar.date(from: DateComponents(year: year, month: month, day: d))!
            let lunar = LunarCalendarUtil.convertSolar2Lunar(dd: d, mm: month, yy: year)
            let lunStr = lunar.lunarDay == 1 ? "\(lunar.lunarDay)/\(lunar.lunarMonth)" : "\(lunar.lunarDay)"
            let wd = calendar.component(.weekday, from: date)
            let sHol = HolidayUtil.getSolarHoliday(dd: d, mm: month)
            let lHol = HolidayUtil.getLunarHoliday(lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth)
            let hasEvent = lunar.lunarDay == 1 || lunar.lunarDay == 15 || sHol != nil || lHol != nil
            let isToday = todayComponents.year == year && todayComponents.month == month && todayComponents.day == d

            let jd = LunarCalendarUtil.jdFromDate(dd: d, mm: month, yy: year)
            let truc = TrucNgayCalculator.getTrucNgay(jd: jd, lunarMonth: lunar.lunarMonth)
            let sao = SaoChieuCalculator.getSaoChieu(jd: jd)
            let act = DayActivityCalculator.getDayActivities(jd: jd, lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth)
            let rating = calculateDayRating(truc: truc, sao: sao, activities: act)

            days.append(CalendarDay(
                solarDay: d, solarMonth: month, solarYear: year,
                lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth,
                isCurrentMonth: true,
                isToday: isToday,
                isSunday: wd == 1,
                isSaturday: wd == 7,
                isHoliday: sHol != nil || lHol != nil,
                hasEvent: hasEvent,
                lunarDisplayText: lunStr,
                dayRatingLabel: rating.label
            ))
        }

        // Next month days
        let totalCells = offset + daysInMonth
        let remaining = totalCells % 7 == 0 ? 0 : 7 - (totalCells % 7)
        let nextMonth = calendar.date(byAdding: .month, value: 1, to: firstDay)!
        let nextMonthValue = calendar.component(.month, from: nextMonth)
        let nextYearValue = calendar.component(.year, from: nextMonth)

        for d in 1...max(remaining, 1) {
            if d > remaining { break }
            let lunar = LunarCalendarUtil.convertSolar2Lunar(dd: d, mm: nextMonthValue, yy: nextYearValue)
            let lunStr = lunar.lunarDay == 1 ? "\(lunar.lunarDay)/\(lunar.lunarMonth)" : "\(lunar.lunarDay)"
            let date = calendar.date(from: DateComponents(year: nextYearValue, month: nextMonthValue, day: d))!
            let wd = calendar.component(.weekday, from: date)
            days.append(CalendarDay(
                solarDay: d, solarMonth: nextMonthValue, solarYear: nextYearValue,
                lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth,
                isCurrentMonth: false, isToday: false,
                isSunday: wd == 1, isSaturday: wd == 7,
                isHoliday: false, hasEvent: false, lunarDisplayText: lunStr
            ))
        }

        return days
    }

    func getUpcomingEvents(dd: Int, mm: Int, yy: Int) -> [UpcomingEvent] {
        var events: [UpcomingEvent] = []
        let todayJd = LunarCalendarUtil.jdFromDate(dd: dd, mm: mm, yy: yy)

        for i in 0..<14 {
            let checkJd = todayJd + i
            let (cd, cm, _) = LunarCalendarUtil.jdToDate(jd: checkJd)
            let lunar = LunarCalendarUtil.convertSolar2Lunar(dd: cd, mm: cm, yy: yy)
            let timeStr: String
            switch i {
            case 0: timeStr = "Hôm nay"
            case 1: timeStr = "Ngày mai · \(cd)/\(cm)"
            default: timeStr = "\(cd)/\(cm)"
            }

            if lunar.lunarDay == 15 {
                events.append(UpcomingEvent(title: "Rằm tháng \(lunar.lunarMonth) Âm lịch", timeLabel: timeStr, tag: "Âm lịch", colorType: .teal))
            }
            if lunar.lunarDay == 1 {
                events.append(UpcomingEvent(title: "Mùng 1 tháng \(lunar.lunarMonth) Âm lịch", timeLabel: timeStr, tag: "Âm lịch", colorType: .teal))
            }
            if let sHol = HolidayUtil.getSolarHoliday(dd: cd, mm: cm) {
                events.append(UpcomingEvent(title: sHol, timeLabel: timeStr, tag: "Ngày lễ", colorType: .red))
            }
            if let lHol = HolidayUtil.getLunarHoliday(lunarDay: lunar.lunarDay, lunarMonth: lunar.lunarMonth),
               lunar.lunarDay != 15, lunar.lunarDay != 1 {
                events.append(UpcomingEvent(title: lHol, timeLabel: timeStr, tag: "Âm lịch", colorType: .gold))
            }

            if events.count >= 5 { break }
        }

        // Add upcoming solar term
        let tkInfo = TietKhiCalculator.getCurrentSolarTerm(dd: dd, mm: mm, yy: yy)
        if let nd = tkInfo.next, tkInfo.daysUntilNext <= 14 {
            let timeStr: String
            switch tkInfo.daysUntilNext {
            case 0: timeStr = "Hôm nay"
            case 1: timeStr = "Ngày mai · \(nd.dd)/\(nd.mm)"
            default: timeStr = "\(nd.dd)/\(nd.mm) · Bắt đầu tiết khí mới"
            }
            events.append(UpcomingEvent(title: "Tiết \(nd.name)", timeLabel: timeStr, tag: "Tiết khí", colorType: .red))
        }

        return Array(events.prefix(4))
    }

    // MARK: - Private Helpers

    private func getHourCanChi(jd: Int, hour: Int) -> String {
        let can = CanChiCalculator.THIEN_CAN
        let chi = CanChiCalculator.DIA_CHI
        let chiIdx: Int
        switch hour {
        case 23, 0: chiIdx = 0
        case 1, 2: chiIdx = 1
        case 3, 4: chiIdx = 2
        case 5, 6: chiIdx = 3
        case 7, 8: chiIdx = 4
        case 9, 10: chiIdx = 5
        case 11, 12: chiIdx = 6
        case 13, 14: chiIdx = 7
        case 15, 16: chiIdx = 8
        case 17, 18: chiIdx = 9
        case 19, 20: chiIdx = 10
        case 21, 22: chiIdx = 11
        default: chiIdx = 0
        }
        let dayCan = ((jd + 9) % 10 + 10) % 10
        let canIdx = (dayCan * 2 + chiIdx) % 10
        return "\(can[canIdx]) \(chi[chiIdx])"
    }

    private func calculateDayRating(truc: TrucNgayInfo, sao: SaoChieuInfo, activities: DayActivities) -> DayRatingInfo {
        var score = 50

        switch truc.rating {
        case "Tốt": score += 20
        case "Xấu": score -= 15
        default: break
        }

        switch sao.rating {
        case "Tốt": score += 20
        case "Xấu": score -= 15
        default: break
        }

        if activities.isNguyetKy { score -= 15 }
        if activities.isTamNuong { score -= 10 }
        score += min(activities.nenLam.count * 2, 10)
        score = max(10, min(100, score))

        let label: String
        switch score {
        case 80...: label = "Rất tốt"
        case 60...: label = "Tốt"
        case 40...: label = "Trung bình"
        default: label = "Xấu"
        }

        return DayRatingInfo(label: label, percent: score)
    }
}
